### PR TITLE
FIP-31 (formerly FIP-30b) DAO script

### PR DESCRIPTION
**Pull request by Joeysantoro**
Upgrade script for transitioning to new FeiDAO timelock. Does not revoke guardian from old timelock to preserve rollback capability.

TODO:
- [x] e2e scropt
- [x] maximize Kashi withdrawal

### CODE REVIEW

File: proposals/dao/fip_30b.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ 6. Transfer 1M kashi DPI LP tokens to OA Timelock
+ 7. Transfer 2.5M kashi ETH LP tokens to OA Timelock
+ 8. Transfer 2.5M kashi TRIBE LP tokens to OA Timelock
+ 9. Transfer 2.5M kashi xSUSHI LP tokens to OA Timelock
+*/
+async function run(addresses, oldContracts, contracts, logging = false) {}
+
+async function teardown(addresses, oldContracts, contracts, logging) {}
+
+async function validate(addresses, oldContracts, contracts) {
```
Comment (ditchfieldcaleb) [inline proposals/dao/fip_30b.ts:21]: type these functions appropriately, eg ValidateUpgradeFunc

Comment (Joeysantoro) [reply proposals/dao/fip_30b.ts:23]: done

File: proposals/dao/feiRariStaking.ts [hunk 2]
```diff
@@ -19,7 +19,7 @@ Steps:
 */
 
 const setup: SetupUpgradeFunc = async (addresses, oldContracts, contracts, logging) => {
-  await increaseTime(5000000);
+  await increaseTime(500000);
```
Comment (ditchfieldcaleb) [inline proposals/dao/feiRariStaking.ts:22]: Why this change?

File: proposals/dao/fip_30b.ts (hunk 1 above)
Comment (ditchfieldcaleb) [reply proposals/dao/fip_30b.ts:23]: you forgot to type validate

File: proposals/dao/fip_30b.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  );
+  expect((await kashiFeiTribe.balanceOf(optimisticTimelock)).toString()).to.be.equal(
+    ethers.constants.WeiPerEther.mul(2_500_000).toString()
+  );
+  expect((await kashiFeiXSushi.balanceOf(optimisticTimelock)).toString()).to.be.equal(
+    ethers.constants.WeiPerEther.mul(2_500_000).toString()
+  );
+}
+
+module.exports = {
```
Comment (ditchfieldcaleb) [inline proposals/dao/fip_30b.ts:50]: I'm going to keep complaining about module.exports every time I see it lol

File: test/integration/e2e.spec.ts [hunk 4]
```diff
@@ -56,6 +55,32 @@ describe('e2e', function () {
...
+      const eta = (await latestTime()) + 100000;
+      const timelockSigner = await getImpersonatedSigner(feiDAO.address);
+      const q = await feiDAOTimelock.connect(timelockSigner).queueTransaction(deployAddress, 100, '', '0x', eta);
+
+      const txHash = (await q.wait()).events[0].args[0];
+      expect(await feiDAOTimelock.queuedTransactions(txHash)).to.be.equal(true);
+
+      await feiDAOTimelock
+        .connect(await getImpersonatedSigner(deployAddress))
+        .vetoTransactions([deployAddress], [100], [''], ['0x'], [eta]);
```
Comment (ditchfieldcaleb) [inline test/integration/e2e.spec.ts:71]: is deployAddress the guardian? if not, how does this work?

File: test/integration/e2e.spec.ts [hunk 5]
```diff
@@ -56,6 +55,32 @@ describe('e2e', function () {
...
+      const { feiDAO, feiDAOTimelock, timelock } = contracts;
+
+      expect(await feiDAO.timelock()).to.be.equal(feiDAOTimelock.address);
+      await feiDAOTimelock.connect(await getImpersonatedSigner(contractAddresses.multisig)).rollback();
+      expect(await feiDAO.timelock()).to.be.equal(timelock.address);
+    });
+  });
+
   describe('Fei DAO', function () {
     it.skip('rollback succeeds', async function () {
```
Comment (ditchfieldcaleb) [inline test/integration/e2e.spec.ts:85]: remove this skip

**Review by ditchfieldcaleb — CHANGES_REQUESTED** [review]
Should add an integration test that rolls back to the old timelock and then executes any random action.

File: proposals/dao/fip_30b.ts (hunk 3 above)
Comment (Joeysantoro) [reply proposals/dao/fip_30b.ts:50]: Eventually there will only be good code to copy

File: proposals/description/fip_30b.txt [hunk 6]
```diff
@@ -1 +1,19 @@
-text
\ No newline at end of file
+Summary:
+
+FEI DAO Timelock Upgrade
+
+Upgrades the FEI DAO Timelock and grants the guardian the power to veto proposals.
+Adds in functionality to roll back to the old timelock via DAO vote with a deadline of November 1st, 2021.
```
Comment (Joeysantoro) [inline proposals/description/fip_30b.txt]: Can take this line out of summary. Also the title is automatically appended from the json so can remove the title line as well

File: scripts/utils/constructProposal.ts [hunk 7]
```diff
@@ -17,6 +17,10 @@ export default async function constructProposal(proposalName: string, logging =
   const contracts = await getAllContracts();
   const proposalBuilder = proposals.builders.alpha();
 
+  if (proposalName === 'fip_31') {
+    proposalBuilder.setGovernor('0x0BEF27FEB58e857046d630B2c03dFb7bae567494');
```
Comment (ElliotFriedman) [inline scripts/utils/constructProposal.ts:21]: Please remove this

File: proposals/description/fip_31.json [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+            "values": "0",
+            "method": "transfer(address,uint256)",
+            "arguments": [
+                "0xbC9C084a12678ef5B516561df902fdc426d95483",
+                "2500000000000000000000000"
+            ],
+            "description": "Transfer 2.5M kashi xSUSHI LP tokens to OA Timelock"
+        }
+    ]
+}
```
Comment (ElliotFriedman) [inline proposals/description/fip_31.json:99]: Double checked this and it looks good!

**Review by ditchfieldcaleb — APPROVED** [review]
